doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,303 @@
+describes a class used to create a descriptor, which provides the methods
+to convert the strided memory in an ndarray into objects that can be
+manipulated, printed, and used in loops. In this NEP, **dtype** is used as the
+name of the class, and **descriptor** is used as the name of the instance. A
+descriptor is what is held by a ndarray object, in addition to a block of
+memory and information about the strides and shape of elements in that memory.
+Creating an instance of ``dtype`` *i.e.* ``a = np.dtype('int8')`` will result
+in a descriptor, which is a python object of type ``dtype``.
+
+The ``dtype`` obect instance ``a`` has attributes, among them ``a.type``, which
+is a class object. Instantiating that class object ``a.type(3)`` produces a
+Numpy `scalar <http://www.numpy.org/devdocs/reference/arrays.scalars.html>`_.
+
+This NEP proposes a class hierarchy for dtypes. The ``np.dtype`` class will
+become an abstract base class, and a number of new classes will be created with
+a hierarchy like scalars. They will support subclassing. A future NEP may
+propose unifying the scalar and dtype type systems, but that is not a goal of
+this NEP.  The changed dtype will:
+
+- facilitate extending dtypes, typically for things like categoricals, novel
+  representations like datetime or IP addresses, or adding attributes like
+  units directly on the dtypes.
+- Allow replacement of the switch-type code we have in places like arrayprint_,
+  PyArray_AdaptFlexibleDtype_, and PyArray_CanCastSafely_ (just a few of many),
+  with method-based overrides, simplifying internal logic and allowing
+  subclassed dtypes to override the functionality.
+
+.. _arrayprint: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/arrayprint.py#L418
+.. _PyArray_AdaptFlexibleDtype: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/src/multiarray/convert_datatype.c#L164 
+.. _PyArray_CanCastSafely: https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/src/multiarray/convert_datatype.c#L398
+
+Overall Design
+--------------
+
+The ``Dtype`` class (and any subclass without ``itemsize``) is effectively an
+abstract base class, as it cannot be used to create instances. A class
+hierarchy is used to add datatype-specific functionality such as ``names`` and
+``fields`` for structured dtypes. The current behaviours are preserved:
+
+- ``np.dtype(obj, align=False, copy=False)`` calls ``arraydescr_new`` with
+  various types of ``obj``:
+  - ``int``, ``np.genenric`` scalar classes, list or dict are parsed into
+    appropriate dtypes
+- singletons are returned where appropriate
+
+Additionally, dtype subclasses are passed through to the subclass ``__new__``
+
+A prototype without error checking, without options handling, and describing
+only ``np.dtype(np.uint8)`` together with an overridable ``get_format_function``
+for ``arrayprint`` looks like::
+
+    import numpy as np
+
+    class Dtype():
+        def __new__(cls, *args, **kwargs):
```

Comment:
Thinking more, the metaclass does in fact seem the best idea, but it should check for the presence of `itemsize` on any *instances* not on the class, i.e., it should let `type` do its work, and then check that `itemsize` exists. This, of course, is just what `ABCmeta` does...